### PR TITLE
Feature/single label task

### PR DESCRIPTION
**Pull request by vesteinn**

### CODE REVIEW

File: src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+)
+from greynirseq.nicenlp.data.encoding import get_word_beginnings
+from greynirseq.nicenlp.utils.label_schema.label_schema import (
+    label_schema_as_dictionary,
+    parse_label_schema,
+    make_dict_idx_to_vec_idx,
+    make_vec_idx_to_dict_idx,
+    make_group_name_to_group_attr_vec_idxs,
+    make_group_masks,
+)
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py:45]: There seem to be many unused imports.

File: src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        self.is_word_initial = is_word_initial
+        self.num_labels = len(self._label_dictionary.symbols)
+
+    def add_args(parser):
+        """Add task-specific arguments to the parser."""
+        parser.add_argument("data", metavar="FILE", help="file prefix for data")
+      
+    @classmethod
+    def setup_task(cls, args: argparse.Namespace, **kwargs):
+        data_dict = Dictionary.load(os.path.join(args.data, "dict.txt"))
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py:80]: Hardcoded `dict.txt`, maybe we should provide this as an option with default value `dict.txt`

File: src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+
+@register_task("multi_class_token_classification_task")
+class MultiClassTokenClassificationTask(FairseqTask):
+    def __init__(
+        self,
+        args: argparse.Namespace,
+        data_dictionary: Dictionary,
+        label_dictionary: Dictionary,
+        is_word_initial: torch.Tensor,
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py:60]: Is the type Tensor or Dict?

File: src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        dataset = {
+            "id": IdDataset(),
+            "net_input": {
+                "src_tokens": RightPadDataset(
+                    src_tokens, pad_idx=self.source_dictionary.pad()
+                ),
+                "nsrc_tokens": NumelDataset(src_tokens),
+                "word_mask": RightPadDataset(word_mask, pad_idx=0),
+            },
+            "target_attrs": RightPadDataset(clean_labels, pad_idx=1),
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py:126]: Could you provide a comment about why the pad_idx are different?

File: src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        src_tokens = data_utils.load_indexed_dataset(
+            str(inputs_path),
+            self.source_dictionary,
+            self.args.dataset_impl,
+            combine=combine,
+        )
+        assert src_tokens is not None, "could not find dataset: {}".format(inputs_path)
+  
+        with data_utils.numpy_seed(self.args.seed):
+            shuffle = np.random.permutation(len(src_tokens))
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py:101]: What is happening with the shuffle? Could you add a comment? The shuffle seems to be overwritten a few lines below.

File: src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+                src_tokens, self.dictionary, self.is_word_initial
+            ),
+            "nsentences": NumSamplesDataset(),
+        }
+        dataset = NestedDictionaryDatasetFix(dataset, sizes=[src_tokens.sizes])
+        return dataset
+
+    def prepare_sentences(self, sentences: List[str]):
+        tokens = [
+            self.encode(token_utils.tokenize_to_string(sentence))
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py:169]: Where is `encode` defined?

File: src/greynirseq/nicenlp/models/multiclass.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+from fairseq import utils
+from fairseq.models.roberta.model import base_architecture, roberta_base_architecture, roberta_large_architecture, RobertaModel
+from fairseq.models.roberta.hub_interface import RobertaHubInterface
+from fairseq.models.roberta.model import RobertaEncoder
+from fairseq.models import register_model, register_model_architecture
+from fairseq.modules import LayerNorm
+
+from greynirseq.nicenlp.utils.label_schema.label_schema import make_vec_idx_to_dict_idx
+
+from greynirseq.nicenlp.utils.constituency import token_utils
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/models/multiclass.py:22]: A few unused imports

File: src/greynirseq/nicenlp/data/datasets.py [hunk 8]
```diff
@@ -237,6 +237,25 @@ def __getitem__(self, index: int):
...
+    ):
+        super().__init__(dataset)
+        self.dictionary = dictionary
+        self.has_bos = True if dataset[0][0] == self.dictionary.bos() else False
+        self.has_eos = True if dataset[0][-1] == self.dictionary.eos() else False
+
+    def __getitem__(self, index: int):
+        item = self.dataset[index]
+        start = 1 if self.has_bos else 0
+        if self.has_eos:
```
Comment (peturorri) [inline src/greynirseq/nicenlp/data/datasets.py:254]: Þetta myndi virka líka:
    end = -1 if self.has_bos else None
    return item[start:end]
(þá er start og end meðhöndlað eins)

File: src/greynirseq/utils/preprocessing/parse_ifd.py [hunk 9]
```diff
@@ -21,16 +27,23 @@ def load_file(
             if inp is not None:
                 o_d.writelines("{}\n".format(" ".join(inp)))
             if lab is not None:
-                o_l.writelines(
-                    "{}\n".format(" <SEP> ".join([" ".join(slab) for slab in lab]))
-                )
+                if sep != " ":
+                    o_l.writelines(
+                        "{}\n".format(sep.join([" ".join(slab) for slab in lab]))
```
Comment (peturorri) [inline src/greynirseq/utils/preprocessing/parse_ifd.py:32]: Mér finnst þetta vera óþarflega flókinn oneliner.

File: src/greynirseq/utils/preprocessing/parse_ifd.py [hunk 10]
```diff
@@ -21,16 +27,23 @@ def load_file(
...
+                else:
+                    o_l.writelines("{}\n".format(sep.join(lab)))
             inp = []
             lab = []
             while line and line != "\n":
                 line = line.strip().split()
                 if len(line):
                     inp.append(line[0])
-                    lab.append(ifd2labels(line[1]))
+                    # Hard-coded for NER..
```
Comment (peturorri) [inline src/greynirseq/utils/preprocessing/parse_ifd.py:42]: :'(

File: src/greynirseq/nicenlp/models/multiclass.py [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+        attr_logits = self.out_proj(x)
+        return attr_logits
+
+
+@register_model("multiclass_roberta")
+class MutliClassRobertaModel(RobertaModel):
+    def __init__(self, args, encoder, task):
+        super().__init__(args, encoder)
+
+        def freeze_module_params(m):
```
[the remark this answers is not in the document]
Comment (vesteinn) [reply src/greynirseq/nicenlp/models/multiclass.py:62]: This is a very simple helper function copied from fairseq.modules.transformer_sentence_encoder.TransformerSentenceEncoder that mutates the model, guessing that's why its there -- but you're right, let's move it.

[the remark this answers is not in the document]
Comment (vesteinn) [reply src/greynirseq/nicenlp/models/multiclass.py:62]: On second thought I'm quite happy to very explicit in the model and it's such a basic function

File: src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py (hunk 2 above)
Comment (vesteinn) [reply src/greynirseq/nicenlp/tasks/multiclass_token_classification_task.py:80]: This is the hard-coding of the default value, could live on the class but is in line with fairseq.

**Review by HaukurPall — APPROVED** [review]
Glæsilegt!
